### DISCUSSION TITLE
Recommended path to detect Uno for source generators

### DISCUSSION THREAD

**Discussion by HavenDV** [unoplatform/uno#9145] [Q&A] [answered]
I'm looking for a way to add automatic detection of Uno.UI and Uno.WinUI projects in my source generator - https://github.com/HavenDV/DependencyPropertyGenerator.
Although Uno has MSBuild specific properties, as far as I know they are the same for Uno.UI and Uno.WinUI (although the HAS_UNO_WINUI constant seems to be present in Uno.WinUI).
In any case, I would like to know a stable way to do this, and not just pick up the right properties that can change.

Current platform detection code:
https://github.com/HavenDV/H.Generators.Extensions#analyzerconfigoptionsproviderextensions

**Comment by jeromelaban** [comment, marked as answer]
`HAS_UNO_WINUI` will stay, we're not going to be changing this any time soon, too much external code depends on it :)

**Reply by HavenDV** [reply]
Thanks for the answer. Is there any plan to add HAS_UNO to the Uno.UI package? Or can I use a non-empty UnoUIUseRoslynSourceGenerators value as an Uno platform indicator?

**Reply by jeromelaban** [reply]
`HAS_UNO` is part of the Uno.UI package as well. It's used by the dotnet templates.

**Reply by HavenDV** [reply]
I create a DefineConstants property for the generator like this (taken from Uno Source Generator):
```xml
<Target Name="CreateDefineConstants" BeforeTargets="GenerateMSBuildEditorConfigFileShouldRun;GenerateMSBuildEditorConfigFileCore">

    <PropertyGroup>
      <DependencyPropertyGenerator_DefineConstants>$(DefineConstants.Replace(';',','))</DependencyPropertyGenerator_DefineConstants>
    </PropertyGroup>

  </Target>
```
As I understand it, the generator incorrectly perceives properties containing ";", so using the original DefineConstants is not possible.

But this method does not see the constants declared by the Uno package (as I understand it, this happens right before the code is compiled, after the generators have been started):
https://github.com/unoplatform/uno/blob/336e8c3ea7c27543d68fd1dc6f8636f9e89abc24/build/uno.winui.common.targets#L19-L46

Can I provide a PR that moves the declaration of Uno constants to an earlier stage?

**Reply by jeromelaban** [reply]
Sure!
